buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -565,10 +566,59 @@ static ByteBuf encodeString0(ByteBufAllocator alloc, boolean enforceHeap, CharBu
         }
     }
 
+    public static String readUtf8(ByteBuf buf, int readerIndex, int length) {
+        if (buf instanceof AbstractByteBuf) {
+            return readUtf8Fast((AbstractByteBuf) buf, readerIndex, length);
+        } else {
+            return readUtf8Slow(buf, readerIndex, length);
+        }
+    }
+
+    private static String readUtf8Slow(ByteBuf buf, int readerIndex, int length) {
+        if (buf instanceof WrappedByteBuf) {
+            ByteBuf wrapped;
+            while ((wrapped = buf.unwrap()) != null) {
+                buf = wrapped;
+            }
+            if (buf instanceof AbstractByteBuf) {
+                return readUtf8Fast((AbstractByteBuf) buf, readerIndex, length);
+            }
+        }
+        UTF8Decoder.UTF8Processor processor = new UTF8Decoder.UTF8Processor(new char[length]);
+        buf.forEachByte(processor);
+        return processor.toString();
+    }
+
+    // Fast-Path for a AbstractByteBuf
+    static String readUtf8Fast(AbstractByteBuf buf, int readerIndex, int length) {
+        buf.checkIndex(readerIndex, length);
+        UTF8Decoder.UTF8Processor processor = new UTF8Decoder.UTF8Processor(new char[length]);
+        /*
+         * Inline AbstractByteBuf.forEachByte by hand since the JIT doesn't do it it properly
```

Comment:
can you provide some background for this? What JIT are you using and how is it configured? Are the thresholds being tripped for this method? If this is a problem we should try to solve it in `AbstractByteBuf` as it should provide a boost to other places as well (as a followup PR).